<a href="https://colab.research.google.com/github/Reis25/CienciaDosDados/blob/master/Lista2_AM_DemetriosReis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Lista 2**
Aprendizado de Máquina

Profs. Evandro Costa e Xu Yang

Aluno: Demétrios Reis Costa

Considere os dados USArrests.csv.

(1) Calcular a matriz covariância S, autovalor e autovetor de S.




In [ ]:
import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import numpy.linalg as linalg

os.getcwd()

file = 'USArrests.csv'

df = pd.read_csv(file)

In [ ]:
# Matriz de covariancia
df.cov()

Murder      Assault    UrbanPop        Rape
Murder     18.970465   291.062367    4.386204   22.991412
Assault   291.062367  6945.165714  312.275102  519.269061
UrbanPop    4.386204   312.275102  209.518776   55.768082
Rape       22.991412   519.269061   55.768082   87.729159

In [ ]:
# Equações
S = df.cov()
S_eigenvalues, S_eigenvectors = np.linalg.eig(S)

In [ ]:
# Auto Valor
S_eigenvalues

array([7.01111485e+03, 2.01992366e+02, 6.16424618e+00, 4.21126508e+01])

In [ ]:
# Auto Vetor
S_eigenvectors

array([[ 0.04170432,  0.04482166, -0.99492173,  0.07989066],
       [ 0.99522128,  0.05876003,  0.0389383 , -0.06756974],
       [ 0.04633575, -0.97685748, -0.05816914, -0.20054629],
       [ 0.0751555 , -0.20071807,  0.07232502,  0.97408059]])

In [ ]:
S_eigenvectors.shape

(4, 4)

(2) Calcular a matriz correlação R, autovalor e autovetor de R.

In [ ]:
# Matriz R de correlação
df.corr()

Murder   Assault  UrbanPop      Rape
Murder    1.000000  0.801873  0.069573  0.563579
Assault   0.801873  1.000000  0.258872  0.665241
UrbanPop  0.069573  0.258872  1.000000  0.411341
Rape      0.563579  0.665241  0.411341  1.000000

In [ ]:
# Equqações: 
R = df.corr()
R_eigenvalues, R_eigenvectors = np.linalg.eig(S)

In [ ]:
# Auto Valor(R):
R_eigenvalues

array([7.01111485e+03, 2.01992366e+02, 6.16424618e+00, 4.21126508e+01])

In [ ]:
# Auto Valor (R)
R_eigenvectors

array([[ 0.04170432,  0.04482166, -0.99492173,  0.07989066],
       [ 0.99522128,  0.05876003,  0.0389383 , -0.06756974],
       [ 0.04633575, -0.97685748, -0.05816914, -0.20054629],
       [ 0.0751555 , -0.20071807,  0.07232502,  0.97408059]])

In [ ]:
# ordenando em ordem descrecente
pares_de_autos = [
    (np.abs(S_eigenvalues[i]),
        S_eigenvectors[:,i]
    ) for i in range(len(S_eigenvalues))
]
pares_de_autos.sort()
pares_de_autos.reverse()

In [ ]:
# calculando a variância explicada e a variância explicada cumulativa
total = sum( S_eigenvalues)
var_exp = [
    (i / total)*100 for i in sorted(
         S_eigenvalues, reverse=True
    )
]
cum_var_exp = np.cumsum(var_exp)

(3) Calcular a contribuição de cada componente principal.

In [ ]:
# visualizando as informações
x = [
    'PC %s' %i for i in range(
        1,len( S_eigenvalues)+1
    )
]
df_temp = pd.DataFrame(
    {'auto-valores':  S_eigenvalues,
     'cum_var_exp':cum_var_exp,
     'var_exp':var_exp,
     'Componente':x}
)
print(df_temp)
print()
print("Auto-vetores")
for autovetor in [p[1] for p in pares_de_autos]:
    print(autovetor)
print()

   auto-valores  cum_var_exp    var_exp Componente
0   7011.114851    96.553422  96.553422       PC 1
1    201.992366    99.335156   2.781734       PC 2
2      6.164246    99.915109   0.579953       PC 3
3     42.112651   100.000000   0.084891       PC 4

Auto-vetores
[0.04170432 0.99522128 0.04633575 0.0751555 ]
[ 0.04482166  0.05876003 -0.97685748 -0.20071807]
[ 0.07989066 -0.06756974 -0.20054629  0.97408059]
[-0.99492173  0.0389383  -0.05816914  0.07232502]



Note que a relevância de de cada auto valor é dada pela divisão entre o auto-valor e a soma entre todos eles. assim 0,96 = 96% = 7011.11/(7011.11 + 201.99 + 6.14 +42.11). É importante ressaltar que por definição a proporção da variância total dos dados é na verdae a sua aplicabilidade.  

(4) Encontra o número k tal que as contribuições acumulam 90% ou mais de
proporção da variância total.

Note que PC1 é praticamente 96% da cobertura dos dados, nisso traduzindo o enunciado é perguntado quantas dimensões são necessárias para conhecermos 90% dos dados e do exposto precisamos apenas de k='1', como podemos ver na matriz do resultado anterior.